Mathlib/Order/Birkhoff.lean

```diff
@@ -0,0 +1,228 @@
+
+namespace UpperSet
+variable [SemilatticeInf α] {s : UpperSet α} {a : α}
+
+@[simp] lemma infIrred_Ici (a : α) : InfIrred (Ici a) := by
+  refine' ⟨fun h ↦ Ici_ne_top h.eq_top, fun s t hst ↦ _⟩
+  have := mem_Ici_iff.2 (le_refl a)
+  rw [←hst] at this
+  exact this.imp (fun ha ↦ le_antisymm (le_Ici.2 ha) $ hst.ge.trans inf_le_left) fun ha ↦
+      le_antisymm (le_Ici.2 ha) $ hst.ge.trans inf_le_right
+
+variable [Finite α]
+
+@[simp] protected lemma infIrred : InfIrred s ↔ ∃ a, Ici a = s := by
+  refine' ⟨fun hs ↦ _, _⟩
+  · obtain ⟨a, ha, has⟩ := (s : Set α).toFinite.exists_minimal_wrt id _ (coe_nonempty.2 hs.ne_top)
+    exact ⟨a, (hs.2 $ erase_inf_Ici ha $ by simpa [eq_comm] using has).resolve_left
+      (lt_erase.2 ha).ne'⟩
+  · rintro ⟨a, rfl⟩
+    exact infIrred_Ici _
+
+end UpperSet
+
+namespace LowerSet
+variable [SemilatticeSup α] {s : LowerSet α} {a : α}
+
+@[simp] lemma supIrred_Iic (a : α) : SupIrred (Iic a) := by
+  refine' ⟨fun h ↦ Iic_ne_bot h.eq_bot, fun s t hst ↦ _⟩
+  have := mem_Iic_iff.2 (le_refl a)
+  rw [←hst] at this
+  exact this.imp (fun ha ↦ (le_sup_left.trans_eq hst).antisymm $ Iic_le.2 ha) fun ha ↦
+    (le_sup_right.trans_eq hst).antisymm $ Iic_le.2 ha
+
+variable [Finite α]
+
+@[simp] protected lemma supIrred : SupIrred s ↔ ∃ a, Iic a = s := by
+  refine' ⟨fun hs ↦ _, _⟩
+  · obtain ⟨a, ha, has⟩ := (s : Set α).toFinite.exists_maximal_wrt id _ (coe_nonempty.2 hs.ne_bot)
+    exact ⟨a, (hs.2 $ erase_sup_Iic ha $ by simpa [eq_comm] using has).resolve_left
+      (erase_lt.2 ha).ne⟩
+  · rintro ⟨a, rfl⟩
+    exact supIrred_Iic _
+
+end LowerSet
+
+section DistribLattice
+variable [DistribLattice α]
+
+section Fintype
+variable [Fintype α] [OrderBot α]
+
+open scoped Classical
+
+/-- **Birkhoff's Representation Theorem**. Any nonempty finite distributive lattice is isomorphic to
+the lattice of lower sets of its sup-irreducible elements. -/
+noncomputable def OrderIso.lowerSetSupIrred : α ≃o LowerSet {a : α // SupIrred a} :=
+  Equiv.toOrderIso
+    { toFun := fun a ↦ ⟨{b | ↑b ≤ a}, fun b c hcb hba ↦ hba.trans' hcb⟩
+      invFun := fun s ↦ (s : Set {a : α // SupIrred a}).toFinset.sup (↑)
+      left_inv := fun a ↦ by
+        refine' le_antisymm (Finset.sup_le fun b ↦ Set.mem_toFinset.1) _
+        obtain ⟨s, rfl, hs⟩ := exists_supIrred_decomposition a
+        exact Finset.sup_le fun i hi ↦
+          le_sup_of_le (b := ⟨i, hs hi⟩) (Set.mem_toFinset.2 $ le_sup (f := id) hi) le_rfl
+      right_inv := fun s ↦ by
+        ext a
+        dsimp
+        refine' ⟨fun ha ↦ _, fun ha ↦ _⟩
+        · obtain ⟨i, hi, ha⟩ := a.2.supPrime.le_finset_sup.1 ha
+          exact s.lower ha (Set.mem_toFinset.1 hi)
+        · dsimp
+          exact le_sup (Set.mem_toFinset.2 ha) }
+    (fun b c hbc d ↦ le_trans' hbc) fun s t hst ↦ Finset.sup_mono $ Set.toFinset_mono hst
+
+attribute [-instance] Set.fintypeIic
+
+/-- Any nonempty finite distributive lattice is isomorphic to its lattice of sup-irreducible lower
+sets. -/
```

Comment:
Could you make a bit clearer (in the docstrings) how this is different from `lowerSetSupIrred`?